core/src/Streamly/Internal/FileSystem/PosixPath/FileDir.hs

```diff
@@ -0,0 +1,178 @@
+{-# LANGUAGE TemplateHaskell #-}
+-- For constraints on "append"
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+#if defined(IS_WINDOWS)
+#define OS_NAME Windows
+#define OS_PATH WindowsPath
+#else
+#define OS_NAME Posix
+#define OS_PATH PosixPath
+#endif
+
+-- |
+-- Module      : Streamly.Internal.FileSystem.OS_PATH.FileDir
+-- Copyright   : (c) 2023 Composewell Technologies
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+-- Portability : GHC
+--
+-- This module provides a type safe path append operation by distinguishing
+-- paths between files and directories. Files are represented by the @File
+-- OS_PATH@ type and directories are represented by the @Dir OS_PATH@ type.
+--
+-- This distinction provides safety against appending a path to a file. Append
+-- operation allows appending to only 'Dir' types.
+--
+module Streamly.Internal.FileSystem.OS_PATH.FileDir
+    (
+    -- * Types
+      File (..)
+    , Dir (..)
+    , IsFileDir
+
+    -- * Statically Verified Path Literals
+    -- | Quasiquoters.
+    , dir
+    , file
+
+    -- * Statically Verified Path Strings
+    -- | Template Haskell expression splices.
+    , dirExp
+    , fileExp
+
+    -- * Operations
+    , append
+    )
+where
+
+import Control.Monad.Catch (MonadThrow(..))
+import Language.Haskell.TH (Q, Exp)
+import Language.Haskell.TH.Syntax (lift)
+import Language.Haskell.TH.Quote (QuasiQuoter)
+import Streamly.Internal.Data.Path (IsPath(..), PathException(..))
+import Streamly.Internal.FileSystem.Path.Common (OS(..), mkQ)
+import Streamly.Internal.FileSystem.OS_PATH (OS_PATH(..))
+
+import qualified Streamly.Internal.Data.Array as Array
+import qualified Streamly.Internal.Data.Stream as Stream
+import qualified Streamly.Internal.FileSystem.Path.Common as Common
+import qualified Streamly.Internal.FileSystem.OS_PATH as OsPath
+
+{- $setup
+>>> :m
+>>> :set -XQuasiQuotes
+
+For APIs that have not been released yet.
+
+>>> import Streamly.Internal.FileSystem.PosixPath (PosixPath)
+>>> import Streamly.Internal.FileSystem.PosixPath.FileDir (File, Dir, file, dir)
+>>> import qualified Streamly.Internal.FileSystem.PosixPath as Path
+>>> import qualified Streamly.Internal.FileSystem.PosixPath.FileDir as PathFD
+-}
+
+newtype File a = File a
+newtype Dir a = Dir a
+
+-- | Constraint to check if a type uses 'File' or 'Dir' as the outermost
+-- constructor.
+class IsFileDir a
+
+instance IsFileDir (File a)
+instance IsFileDir (Dir a)
+
+instance IsPath OS_PATH (File OS_PATH) where
+    unsafeFromPath = File
+
+    -- Cannot have "." or ".." as last component.
+    fromPath p@(OS_PATH arr) = do
+        s1 <-
+                Stream.toList
+                    $ Stream.take 3
+                    $ Stream.takeWhile (not . Common.isSeparator OS_NAME)
+                    $ fmap Common.wordToChar
+                    $ Array.readRev arr
+        -- XXX On posix we just need to check last 3 bytes of the array
+        case s1 of
+            '.' : xs ->
```

Comment:
Also, what if this ends with `/`. That way `s1` is an empty list.
Eg, `abc/def/`

Reply:
We parse that as a correct file path?

Reply:
`OS_PATH` cannot end with `/` we strip that. This assumes that we have a valid `OS_PATH`.